[stm32] Fix compilation of SPI DMA driver with GCC 11+

chris-durand opened the pull request:
Fix incorrect access specifier in `SpiMaster` that prevents `SpiMasterDma` from compiling with GCC 11.

GCC versions prior to 11 allowed accessing private member variables of template base classes from sub-classes: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58993

cc @hshose 

Fixes #876

chris-durand committed:
[stm32] Fix compilation of SPI DMA driver with GCC 11+

GCC prior to 11 allowed accessing private member variables of
template base classes from sub-classes:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58993

rleh reviewed (changes requested):
Same fix is also required for other SpiMaster implementations:
- [STM32 UART](https://github.com/modm-io/modm/blob/develop/src/modm/platform/spi/stm32_uart/uart_spi_master.hpp.in#L39)
- [Atmel SAM](https://github.com/modm-io/modm/blob/develop/src/modm/platform/spi/sam/spi_master.hpp.in#L29)
- [Raspberry RP2040](https://github.com/modm-io/modm/blob/develop/src/modm/platform/spi/rp/spi_master.hpp.in#L32)
- Maybe [Atmel AVR](https://github.com/modm-io/modm/blob/develop/src/modm/platform/spi/at90_tiny_mega/spi_master.hpp.in#L55)?

chris-durand commented:
The fix is only required if you derive from the `SpiMaster` class and access that specific member variable from the subclass. Apart from the STM32 SPI driver that same bug only exists in the RP Pico driver. I would not preemptively change it for the others because that `state` variable is not part of the public interface. Whenever someone adds a subclass that requires it they can change it.

chris-durand committed:
[rp2040] Fix compilation of SPI DMA driver with GCC 11+

GCC prior to 11 allowed accessing private member variables of
template base classes from sub-classes:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58993
